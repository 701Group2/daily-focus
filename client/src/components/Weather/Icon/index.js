import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
    faCloud,
    faCloudMoon,
    faCloudMoonRain,
    faCloudSun,
    faCloudSunRain,
    faCloudShowersHeavy,
    faBolt,
    faSnowflake,
    faSun,
    faSmog,
    faMoon,
} from "@fortawesome/free-solid-svg-icons";
import PropTypes from "prop-types";
import styles from "./style.module.scss";

// The icon labels retrieved from the API are matched up with icons
// 'n' means night and 'd' means day
const weatherIcon = {
    // Thunderstorm
    "11d": <FontAwesomeIcon icon={faBolt} className={styles.faColor} />,
    "11n": <FontAwesomeIcon icon={faBolt} className={styles.faColor} />,

    // Rain
    "10d": <FontAwesomeIcon icon={faCloudShowersHeavy} className={styles.faColor} />,
    "10n": <FontAwesomeIcon icon={faCloudShowersHeavy} className={styles.faColor} />,

    // Light Rain/Showers
    "09d": <FontAwesomeIcon icon={faCloudSunRain} className={styles.faColor} />,
    "09n": <FontAwesomeIcon icon={faCloudMoonRain} className={styles.faColor} />,

    // Snow
    "13d": <FontAwesomeIcon icon={faSnowflake} className={styles.faColor} />,
    "13n": <FontAwesomeIcon icon={faSnowflake} className={styles.faColor} />,

    // Atmosphere
    "50d": <FontAwesomeIcon icon={faSmog} className={styles.faColor} />,
    "50n": <FontAwesomeIcon icon={faSmog} className={styles.faColor} />,

    // Clear Sky
    "01d": <FontAwesomeIcon icon={faSun} className={styles.faColor} />,
    "01n": <FontAwesomeIcon icon={faMoon} className={styles.faColor} />,

    // Cloudy
    "04d": <FontAwesomeIcon icon={faCloud} className={styles.faColor} />,
    "04n": <FontAwesomeIcon icon={faCloud} className={styles.faColor} />,

    // Partly Cloudy
    "02d": <FontAwesomeIcon icon={faCloudSun} className={styles.faColor} />,
    "02n": <FontAwesomeIcon icon={faCloudMoon} className={styles.faColor} />,
    "03d": <FontAwesomeIcon icon={faCloudSun} className={styles.faColor} />,
    "03n": <FontAwesomeIcon icon={faCloudMoon} className={styles.faColor} />,
};

function Icon({ icon }) {
    return <div>{weatherIcon[icon]}</div>;
}

Icon.propTypes = {
    icon: PropTypes.string,
};

export default Icon;
