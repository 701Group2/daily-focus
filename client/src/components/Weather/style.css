body{
    background-color: white;
}

.container{
    top: 40px;
    left: 40px;
    margin: auto;
    width: 370px;
    height: 379px;
    border: 3px solid gray;
    padding: 0px;
}

.weather-text-box{
    
    text-align: center;
    background-color: lightblue;
    height: 64px;
}

.weather-text{
    font-size:36px;
    
}



.icon-container{
    float:left;
}

.discription{
    float:right;
}
 