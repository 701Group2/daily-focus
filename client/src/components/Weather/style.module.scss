
.weatherWidget{
    position: relative;
    width: 450px;
    height: 560px;
    
    background-color:gray;
}

.container{
    position: absolute;
    width: 340px;
    height: 399px;
    left: 55px;
    top: 80px;

    background-color: #FFFFFF;
    box-shadow: 0px 4px 16px 8px rgba(0, 0, 0, 0.25);
    border-radius: 25px;   
}

.header{   
    //position: absolute;
    
    width: 340px;
    height: 47.88px;
    left: 55px;
    top: 80px;

    background: #30A0F5;
    border-radius: 25px 25px 0px 0px;
    
}

.weatherText{
    
    font-family: Ropa Sans;
    font-style: normal;
    font-weight: normal;
    font-size: 30px;
    line-height: 32px;
    display: flex;
    align-items: center;
    text-align: center;
    color: #FFFFFF;
    
}

.weatherTextBox{
   
    width: 251px;
    height: 47.88px;
    left: 97px;
    top: 79px;

    align-items: center;
    text-align: center;

}



.iconContainer{
    float:left;
}

.description{
    float:right;
}

.hourlyForecast{
    //position: absolute;
    width: 300px;
    height: 67.83px;
    left: 75px;
    top: 247.58px;
    //margin: auto;
    background-color: white;
    display:flex;
    flex-direction:row;
    justify-content: space-around;

    

}

.dailyForecast{
    //position: absolute;
    width: 291px;
    height: 123.69px;
    left: 75px;
    top: 341.34px;
  
    //margin: auto;
    background-color: white;
    display:flex;
    flex-direction: column;


  

}


.currentWeatherBox{
    //position: absolute;
    width: 300px;
    height: 80.8px;
    left: 75px;
    top: 140.85px;
    background-color: white;
    display: flex;
    justify-content: space-around;


}
 