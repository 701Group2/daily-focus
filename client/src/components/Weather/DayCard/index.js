import Icon from "../Icon";
import styles from "./DayCard.module.scss";

function DayCard(props) {
    var days = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];
    var dayNum = new Date(props.day.dt * 1000).getDay();
    var result = days[dayNum];

    return (
        <div className={styles.card}>
            <div className={styles.day}>{result}</div>
            <div className={styles.iconContainer}>
                <Icon icon={props.day.weather[0].icon} />
            </div>

            <div className={styles.lowTemp}>{Math.floor(props.day.temp.min)}°C</div>
            <div className={styles.highTemp}>{Math.floor(props.day.temp.max)}°C</div>
        </div>
    );
}

export default DayCard;
