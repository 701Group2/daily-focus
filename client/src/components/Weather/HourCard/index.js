import Icon from "../Icon";
import styles from "./style.module.scss";

var moment = require("moment");

function HourCard(props) {
    let dt = props.hour.dt;
    let date = new Date(dt * 1000);
    // Hours part from the timestamp
    let hours = date.getHours();
    let minutes = "0" + date.getMinutes();
    // Seconds part from the timestamp
    let seconds = "0" + date.getSeconds();

    // Will display time in 10:30:23 format
    let formattedTime = hours + ":" + minutes.substr(-2) + ":" + seconds.substr(-2);

    return (
        <div className={styles.hourCard}>
            <div className={styles.time}>{moment(formattedTime, "HH:mm:ss").format("hh:mma")}</div>
            <div className={styles.icon}>
                <Icon icon={props.hour.weather[0].icon} />
            </div>
            <div className={styles.temp}>{Math.floor(props.hour.temp)}Â°C</div>
        </div>
    );
}

export default HourCard;
