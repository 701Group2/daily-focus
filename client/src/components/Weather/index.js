import React, { useState, useEffect } from "react";
import DayCard from "./DayCard/index.js";
import HourCard from "./HourCard/index.js";
import Icon from "./Icon/index.js";
import styles from "./style.module.scss";

const API_KEY = "2d6026fb2c6a6ef3bbbc1f81c56baf04";

const aucklandCoords = { lat: -36.85, lon: 174.76 };

function Weather() {
    const [icon, setIcon] = useState("04d");
    const [main, setMain] = useState(undefined);
    const [temp, setTemp] = useState(undefined);
    const [maxTemp, setMaxTemp] = useState(null);
    const [minTemp, setMinTemp] = useState(null);
    // const [description, setDescription] = useState("");
    const [dailyData, setDailyData] = useState([]);
    const [hourlyData, setHourlyData] = useState([]);
    const [name, setName] = useState(null);
    const [country, setCountry] = useState(null);

    const formatDayCards = (dailyData) => {
        return dailyData.map((day, index) => <DayCard day={day} key={index} />);
    };

    const formatHourCards = (hourlyData) => {
        let hourArray = hourlyData.map((hour, index) => <HourCard hour={hour} key={index} />);
        return hourArray;
    };

    const getWeatherData = (lat, lon) => {
        // Get all the weather data (current, hourly for current day, and daily)
        fetch(
            `https://api.openweathermap.org/data/2.5/onecall?lat=${lat}&lon=${lon}&units=metric&exclude=alerts,minutely&appid=${API_KEY}`
        )
            .then((res) => res.json())
            .then((response) => {
                // Current weather
                setMain(response.current.weather[0].main);
                setTemp(Math.floor(response.current.temp));
                setMaxTemp(Math.floor(response.daily[0].temp.max));
                setMinTemp(Math.floor(response.daily[0].temp.min));
                // setDescription(response.current.weather[0].description);
                setIcon(response.current.weather[0].icon);

                // Hourly weather for next 6 hours
                const hourlyData = response.hourly.slice(0, 6);
                setHourlyData(hourlyData);

                // Daily weather for next 5 days
                const dailyData = response.daily.slice(1, 6);
                setDailyData(dailyData);
            });

        // Get the city and country name for the requested coordinates
        fetch(
            `http://api.openweathermap.org/geo/1.0/reverse?lat=${lat}&lon=${lon}&appid=${API_KEY}`
        )
            .then((res) => res.json())
            .then((response) => {
                setName(response[0].name);
                setCountry(response[0].country);
            });
    };

    useEffect(() => {
        if ("geolocation" in navigator) {
            navigator.geolocation.getCurrentPosition(
                function (position) {
                    const lat = position.coords.latitude;
                    const lon = position.coords.longitude;
                    getWeatherData(lat, lon);
                },
                // Permission not given to access location, so default to Auckland coordinates
                function onError() {
                    getWeatherData(aucklandCoords.lat, aucklandCoords.lon);
                }
            );
        } else {
            getWeatherData(aucklandCoords.lat, aucklandCoords.lon);
        }
    }, []);

    return (
        <div className={styles.weatherWidget}>
            <div className={styles.container}>
                <div className={styles.headerRectangle}>
                    <div className={styles.weatherTextBox}>Weather</div>
                </div>
                <div className={styles.content}>
                    <div className={styles.currentWeatherBox}>
                        <div className={styles.iconContainer}>
                            <Icon icon={icon} />
                        </div>
                        <div className={styles.description}>
                            <div className={styles.weatherNZ}>
                                {main} | {name}, {country}
                            </div>
                            <div className={styles.temp}>{temp}°C</div>
                            <div className={styles.tempBounds}>
                                L:{minTemp}°C | H:{maxTemp}°C
                            </div>
                        </div>
                    </div>
                    <div className={styles.hourlyForecast}>{formatHourCards(hourlyData)}</div>
                    <div className={styles.dailyForecast}>{formatDayCards(dailyData)}</div>
                </div>
            </div>
        </div>
    );
}

export default Weather;
