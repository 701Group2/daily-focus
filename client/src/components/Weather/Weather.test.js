import React from "react";
import { shallow } from "enzyme";
import ShallowRenderer from "react-test-renderer/shallow";

import Weather from "./";

let component;

beforeEach(() => {
    component = shallow(<Weather />);
});

test("it renders correctly", () => {
    const shallowRenderer = new ShallowRenderer();
    const snapshotComponent = shallowRenderer.render(<Weather />);
    expect(snapshotComponent).toMatchSnapshot();
});
