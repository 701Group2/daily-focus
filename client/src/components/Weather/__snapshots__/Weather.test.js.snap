// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<div
  className="weatherWidget"
>
  <div
    className="container"
  >
    <div
      className="wholeRectangle"
    />
    <div
      className="headerRectangle"
    />
    <div
      className="weatherTextBox"
    >
      Weather
    </div>
    <div
      className="dailyForecast"
    />
    <div
      className="hourlyForecast"
    />
    <div
      className="currentWeatherBox"
    >
      <div
        className="iconContainer"
      />
      <div
        className="description"
      >
        <div
          className="weatherNZ"
        >
          
          |Auckland, NZ
        </div>
        <div
          className="temp"
        >
          °C
        </div>
        <div
          className="tempBounds"
        >
          L:
          °C|H:
          °C
        </div>
      </div>
    </div>
  </div>
</div>
`;
